### PR TITLE
Feature generate from existing data

### PR DESCRIPTION
**Pull request by ronanstokes-db**
## Proposed changes

Addition of features to aid in generating of data from existing data sets

## Types of changes

What types of changes does your code introduce to dbldatagen?
_Put an `x` in the boxes that apply_

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Change to tutorials, tests or examples
- [x] Non code change (readme, images or other non-code assets)
- [x] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. 
If you're unsure about any of them, don't hesitate to ask. We're here to help! 
This is simply a reminder of what we are going to look for before merging your code._

- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules
- [x] Submission does not reduce code coverage numbers
- [x] Submission does not increase alerts or messages from LGTM

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you 
did and what alternatives you considered, etc...

### CODE REVIEW

File: docs/source/generating_from_existing_data.rst [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        .withColumn('date_value', 'date', expr='current_date()')
+        .withColumn('binary_value', 'binary',
+                    expr="cast('dbldatagen generated synthetic data' as binary)" )
+    )
+
+
+When generating code from a schema, the code generation process does not have information about the data attributes
+and so the generated code just illustrates some typical generation expressions.
+
+If we supply a source data frame to the generation process, the data analyzer instance will use information derived
```
Comment (howardwu-db) [inline docs/source/generating_from_existing_data.rst:101]: dataframe is one word

File: docs/source/index.rst [hunk 2]
```diff
@@ -30,6 +30,7 @@ As it is installable via `%pip install`, it can also be incorporated in environm
    Generating repeatable data  <repeatable_data_generation>
    Using streaming data <using_streaming_data>
    Generating JSON and structured column data <generating_json_data>
+   Generating synthetic data from existing data <generating_from_existing_data>
    Generating Change Data Capture (CDC) data<generating_cdc_data>
```
Comment (howardwu-db) [inline docs/source/index.rst:34]: put a space between data the "<" character

File: docs/source/index.rst [hunk 3]
```diff
@@ -30,6 +30,7 @@ As it is installable via `%pip install`, it can also be incorporated in environm
    Generating repeatable data  <repeatable_data_generation>
```
Comment (howardwu-db) [inline docs/source/index.rst:30]: extra space between data the "<"

File: docs/source/index.rst [hunk 4]
```diff
@@ -30,6 +30,7 @@ As it is installable via `%pip install`, it can also be incorporated in environm
...
    Using streaming data <using_streaming_data>
    Generating JSON and structured column data <generating_json_data>
+   Generating synthetic data from existing data <generating_from_existing_data>
    Generating Change Data Capture (CDC) data<generating_cdc_data>
    Using multiple tables <multi_table_data>
    Extending text generation  <extending_text_generation>
```
Comment (howardwu-db) [inline docs/source/index.rst:36]: extra space in between generation and "<"

**Review by howardwu-db — CHANGES_REQUESTED** [review]

File: dbldatagen/data_analyzer.py [hunk 5]
```diff
@@ -3,83 +3,60 @@
...
-This module defines the DataAnalyzer class. This is still a work in progress.
+This module defines the ``DataAnalyzer`` class.
+
+This code is experimental and both APIs and code generated is liable to change in future versions.
 """
-from pyspark.sql.types import StructField
-from pyspark.sql.functions import lit
-from pyspark.sql import functions as fns
+from pyspark.sql.types import LongType, FloatType, IntegerType, StringType, DoubleType, BooleanType, ShortType, \
+    TimestampType, DateType, DecimalType, ByteType, BinaryType,  StructType, ArrayType, DataType
```
[the remark this answers is not in the document]
Comment (ronanstokes-db) [reply dbldatagen/data_analyzer.py]: fixed

File: dbldatagen/data_analyzer.py [hunk 6]
```diff
@@ -3,83 +3,60 @@
...
+    _GENERATED_FROM_SCHEMA_COMMENT = strip_margins("""
+                        |# Column definitions are stubs only - modify to generate correct data  
+                        |#""", '|')
 
-    def __init__(self, df, sparkSession=None):
+    def __init__(self, df=None, sparkSession=None):
         """ Constructor:
-        :param df: data frame to analyze
-        :param sparkSession: spark session to use
+        :param df: Data frame to analyze
```
[the remark this answers is not in the document]
Comment (ronanstokes-db) [reply dbldatagen/data_analyzer.py]: fixed

File: dbldatagen/data_analyzer.py [hunk 7]
```diff
@@ -3,83 +3,60 @@
...
 
 class DataAnalyzer:
     """This class is used to analyze an existing data set to assist in generating a test data set with similar
-    characteristics
-
-    .. warning::
-       Not fully implemented.
+    characteristics, and to generate code from existing schemas and data
 
     :param df: Spark data frame to analyze
```
[the remark this answers is not in the document]
Comment (ronanstokes-db) [reply dbldatagen/data_analyzer.py]: fixed

File: docs/source/generating_from_existing_data.rst (hunk 1 above)
Comment (ronanstokes-db) [reply docs/source/generating_from_existing_data.rst:101]: fixed

File: docs/source/index.rst (hunk 2 above)
Comment (ronanstokes-db) [reply docs/source/index.rst]: fixed

File: docs/source/index.rst (hunk 3 above)
Comment (ronanstokes-db) [reply docs/source/index.rst]: fixed

File: docs/source/index.rst (hunk 4 above)
Comment (ronanstokes-db) [reply docs/source/index.rst]: fixed

**Review by howardwu-db — APPROVED** [review]

**Pull request merged**